pkgs/misc/emulators/citra/base.nix

```diff
@@ -0,0 +1,195 @@
+    owner = "mozilla";
+    repo = "cubeb";
+    rev = "1d66483ad2b93f0e00e175f9480c771af90003a7";
+    sha256 = "06rlmdarrk9hwm8svmh2nl2cd4ybmzm2jhp7ih6xzzrba31hg7n3";
+    fetchSubmodules = true;
+  };
+  dynarmic-src = fetchFromGitHub {
+    owner = "citra-emu";
+    repo = "dynarmic";
+    rev = "af0d4a7c18ee90d544866a8cf24e6a0d48d3edc4";
+    sha256 = "1hsl8vdd2y2h2rmhx3bvh4a8l3m33xnsnk5v64jkwdfcl15iyh2a";
+  };
+  lodepng-src = fetchFromGitHub {
+    owner = "lvandeve";
+    repo = "lodepng";
+    rev = "31d9704fdcca0b68fb9656d4764fa0fb60e460c2";
+    sha256 = "1il97cq1rms3078zsyq00acgb75kva1smcrkb8r450p6c1rmwsii";
+  };
+  nihstro-src = fetchFromGitHub {
+    owner = "neobrain";
+    repo = "nihstro";
+    rev = "fd69de1a1b960ec296cc67d32257b0f9e2d89ac6";
+    sha256 = "1lfx92lpszmz1xsvb80a349r3i7g5msxixkb29v1qvi4a8v3ak58";
+  };
+  soundtouch-src = fetchFromGitHub {
+    owner = "citra-emu";
+    repo = "ext-soundtouch";
+    rev = "060181eaf273180d3a7e87349895bd0cb6ccbf4a";
+    sha256 = "0x92a5wb2hcaqzckpdqsggbf863a9kicpa4xz4g4gdyid19n3ci9";
+  };
+  teakra-src = fetchFromGitHub {
+    owner = "wwylele";
+    repo = "teakra";
+    rev = "3e032a73d7e97eb434a053391d95029eebd7e189";
+    sha256 = "1cpxj8jvkj52f3jpy7rsgzs09h8cd3jmx6f2282dh9w2daaga4w8";
+  };
+  xbyak-src = fetchFromGitHub {
+    owner = "herumi";
+    repo = "xbyak";
+    rev = "c306b8e5786eeeb87b8925a8af5c3bf057ff5a90";
+    sha256 = "13l5yyaf67h6s0zlzksah4rybv4xsl1a7939frvwahxch88wfslr";
+  };
+  # Last updated Thu, 28 Oct 2021 12:43:04 GMT from https://api.citra-emu.org/gamedb/
+  compat-src = ./compatibility_list.json;
+in stdenv.mkDerivation {
+  inherit pname version src;
+  cmakeFlags = [
+    "-DBUILD_SHARED_LIBS=OFF"
+    "-DCITRA_USE_BUNDLED_SDL2=OFF"
+    "-DCITRA_USE_BUNDLED_QT=OFF"
+    "-DCITRA_USE_BUNDLED_FFMPEG=OFF"
+    "-DDYNARMIC_ENABLE_CPU_FEATURE_DETECTION=OFF"
+    "-DENABLE_COMPATIBILITY_LIST_DOWNLOAD=OFF" # Sacrifices reproducibility
+    "-DCITRA_ENABLE_COMPATIBILITY_REPORTING=ON"
+    "-DENABLE_FFMPEG_VIDEO_DUMPER=ON"
+    "-DENABLE_QT_TRANSLATION=ON"
+    "-DUSE_SYSTEM_BOOST=ON"
+    "-GNinja"
+    "-Wno-dev"
+    "-DUSE_DISCORD_PRESENCE=${if discordSupport then "ON" else "OFF"}"
+    "-DENABLE_WEB_SERVICE=${if onlineSupport then "ON" else "OFF"}"
+    (if fdkSupport     then "-DENABLE_FDK=ON"           else "-DENABLE_FFMPEG_AUDIO_DECODER=ON")
+  ];
+
+  nativeBuildInputs = [ catch2 cmake ninja pkg-config qttools wrapQtAppsHook
+    enet fmt inih libressl soundtouch zstd discord-rpc
+  ];
+  buildInputs = [ boost173 ffmpeg libusb1 qtbase qtdeclarative qtmultimedia rapidjson SDL2
+  ] ++ optional alsaSupport        alsa-lib
+    ++ optional fdkSupport         fdk_aac
+    ++ optional jackaudioSupport   libjack2
+    ++ optional pulseSupport       libpulseaudio
+    ++ optional sndioSupport       sndio
+    ++ optional udevSupport        udev;
+
+  postPatch = ''
+    # Manually prepare all our externals instead of fetching git submodules.
```

Comment:
I really don't think avoiding submodules here is a good idea. This would tremendously increase complexity of package and burden of maintaining it, IMHO without any added benefits. Maybe I miss something here?

Reply:
I've read over and over again that fetchSubmodules is best avoided, so I avoided it. 

Upsides:
* Faster build times, fetchSubmodules takes awhile!
* More easily overrideable dependencies by users

Downsides:
* Harder to maintain, but @IvarWithoutBones already made a script for updating it. Thanks Ivar!

I'm of course willing to change it to conform to best practices. We can keep the fetchSubmodles and just devendor the packages that already exist in nixpkgs.

I'd like @SuperSandro2000 's input on this, I often see them complaining about submodules.

Reply:
From my other comment:

> I have written a small script which fetches the submodules, and exports the data `fetchFromGitHub` needs as a json file. The idea is to use `lib.importJSON` on that, and extract the result to the location citra expects. It was fun to write, but I think it's just not practical to maintain. The script is currently very unfinished, and I'm not really motivated to work on it. Here's the current (messy) state: [IvarWithoutBones@4e155ef](https://github.com/IvarWithoutBones/nixpkgs/commit/4e155ef7fee11283d5dd15ae975f615c01f51bc4).
> 
> I think a better option would be working with upstream, to properly devendor these. Dynarmic is an dependency of Citra and yuzu (currently a submodule) which takes ages to compile. It recently added an cmake install target[[1](https://github.com/merryhime/dynarmic/pull/652)], adding support for that in upstream yuzu and citra would make testing much faster. Right now maintaining yuzu is kind of a pain for me, since I don't have a very fast computer.
> 
> I hope adding support for it shouldn't be that difficult, though I'm not extremely familair with CMake.